File: 2021/day13/day13.rs
//! [Day 13: Transparent Origami](https://adventofcode.com/2021/day/13)

use aoc::ocr::scan_5x6;

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (i32, String) {
    let n = 2000;
    let mut grid = vec![vec![0i8; n]; n];
    let mut part1 = 0;

    for line in data.lines() {
        if line.is_empty() || line.starts_with("fold") {
            break;
        }
        let (x, y) = line.split_once(',').unwrap();
        let xx = x.parse::<usize>().unwrap();
        let yy = y.parse::<usize>().unwrap();
        grid[yy][xx] = 1;
    }

    for line in data.lines() {
        if !line.starts_with("fold") {
            continue;
        }

        if line.starts_with("fold along x=") {
            let (_, s) = line.split_once('=').unwrap();
            let fold = s.parse::<usize>().unwrap();
            for line in &mut grid {
                for x in 0..fold {
                    if line[fold + 1 + x] == 1 {
                        line[fold - 1 - x] = 1;
                    }
                }
            }
            for line in &mut grid {
                line.resize(fold, 0);
            }
        } else if line.starts_with("fold along y=") {
            let (_, s) = line.split_once('=').unwrap();
            let fold = s.parse::<usize>().unwrap();
            for y in 0..fold {
                for x in 0..grid[0].len() {
                    if grid[fold + 1 + y][x] == 1 {
                        grid[fold - 1 - y][x] = 1;
                    }
                }
            }
            grid.resize(fold, vec![0; grid[0].len()]);
        }

        if part1 == 0 {
            for row in &grid {
                for cell in row {
                    part1 += i32::from(*cell);
                }
            }
        }
    }

    let mut crt = String::new();
    for row in &grid {
        for cell in row {
            crt.push(if *cell == 0 { '.' } else { '#' });
        }
        crt.push('\n');
    }

    (part1, scan_5x6(&crt))
}

File: src/year2021/day13/day13.py
#!/usr/bin/env python3

# Day 13: Transparent Origami
# https://adventofcode.com/2021/day/13

import atexit
import sys
import time
from pathlib import Path

data = open("input.txt" if len(sys.argv) == 1 else sys.argv[1]).read().splitlines()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


sys.path.append(Path(__file__).parent.parent.as_posix())
from ocr.ocr import ocr  # noqa

n = 2000
grid = [["." for _ in range(n)] for _ in range(n)]

for line in data:
    if not line or line.startswith("fold"):
        break
    x, y = map(int, line.split(","))
    grid[y][x] = "#"

# for y, row in enumerate(grid):
#     print("%3d" % y, "".join(str(x) for x in row))

part1 = False

for line in data:
    if not line.startswith("fold"):
        continue

    if line.startswith("fold along x="):
        fold = int(line.split("=")[1])
        for y in range(len(grid)):
            for x in range(fold):
                if grid[y][fold + 1 + x] == "#":
                    grid[y][fold - 1 - x] = "#"
            del grid[y][fold:]

    elif line.startswith("fold along y="):
        fold = int(line.split("=")[1])
        for y in range(fold):
            for x in range(len(grid[0])):
                if grid[fold + 1 + y][x] == "#":
                    grid[fold - 1 - y][x] = "#"

        del grid[fold:]

    if not part1:
        print(sum(1 for row in grid for cell in row if cell == "#"))
        part1 = True

crt = "\n".join("".join(str(x) for x in row) for row in grid)
print(ocr(crt))
